Parallel I/O integration

SeanBryan51 opened the pull request:
Experimental branch for integrating parallel I/O (part of #358). This should not be merged into main as is.


<!-- readthedocs-preview cable start -->
----
📚 Documentation preview 📚: https://cable--641.org.readthedocs.build/en/641/

<!-- readthedocs-preview cable end -->

SeanBryan51 committed:
Add parallel I/O decomp utilities

SeanBryan51 committed:
src/offline/cable_mpi.F90: STOP when MPI is unavailable in mpi_grp_abort

SeanBryan51 committed:
Replace abort with cable_abort

SeanBryan51 committed:
Introduce global structure variables

Global structure variables describe the grid of the global simulation
domain, irrespective of the current MPI rank. These are mland_global,
land_x_global, land_y_global, mp_global and landpt_global.

This change updates static (time invariant) arrays in the met file to
be stored on the global land grid (size mland_global). These variables
are:
- cable_iovars.F90:latitude(mland_global)
- cable_iovars.F90:longitude(mland_global)
- cable_iovars.F90:vegtype_metfile(mland_global, nmetpatches)
- cable_iovars.F90:vegpatch_metfile(mland_global, nmetpatches)
- cable_iovars.F90:soiltype_metfile(mland_global, nmetpatches)
- cable_iovars.F90:PrecipScale(mland_global)
- cable_input.F90:elevation(mland_global)
- cable_input.F90:avPrecip(mland_global)

SeanBryan51 committed:
src/offline/cable_mpi.F90: Add legacy communicator procedure overload for mpi_grp_t constructor

SeanBryan51 committed:
Allow for selection of legacy MPI or parallel I/O at runtime

SeanBryan51 committed:
src/offline/cable_mpi.F90: add mpi_grp_split

SeanBryan51 committed:
Introduce local structure variables

Local structure variables describe the grid information local to the
current MPI rank. These are mp, mland, land_x, land_y,
land_decomp_start, land_decomp_end, patch_decomp_start, patch_decomp_end
and landpt.

SeanBryan51 committed:
Add PIO as an optional dependency

SeanBryan51 committed:
Add parallel I/O abstraction

SeanBryan51 committed:
Temporary commit (to remove): add unit tests

SeanBryan51 committed:
Add data structure for I/O decompositions

SeanBryan51 committed:
Pass in mpi_grp to the serial and master drivers

The MPI master driver is passed a group containing a single process as
the I/O in the master driver is done in serial.

SeanBryan51 commented:
Hi @Whyborn, don't want to scare you with a huge PR 😅, but could I get your thoughts on the commits after **Temporary commit (to remove): add unit tests**. This includes some functionality for handling I/O decompositions which did not exist in Mock CABLE

Whyborn commented on src/offline/cable_io_decomp.F90:
```diff
@@ -0,0 +1,257 @@
+      io_decomp%output_base_patch_rad_real32         => io_decomp%patch_rad_to_land_patch_rad_real32
+      io_decomp%output_base_patch_plantcarbon_real32 => io_decomp%patch_plantcarbon_to_land_patch_plantcarbon_real32
+      io_decomp%output_base_patch_soilcarbon_real32  => io_decomp%patch_soilcarbon_to_land_patch_soilcarbon_real32
+    end if
```
Is this to support output of either gridded or vectorised outputs? Why isn't an `IF - ELSEIF - ELSE`? Where do site runs fall here?

Whyborn commented on src/offline/cable_io_decomp.F90:
```diff
@@ -0,0 +1,257 @@
+
+    logical :: requires_land_output_grid, requires_x_y_output_grid
+
+    mem_shape_land              = [dim_spec_t('land', mland)]
```
Might be a dumb question, but where is the concrete `dim_spec_t` used as source to define these dim specs defined?

Whyborn reviewed (commented):
I looked at the CABLE part of the diffs. If I'm understanding right, the code doesn't yet to PIO output right? Just getting the data structures in place to handle it in future? Personally I'd like the decompositions to be handled by the respective science modules, so that they're more self-contained. The module can expose the likely dimensions of the output e.g. `nlat, nlon, ntiles` etc via `get_nlon(), get_nlat, ...`, but I think having them exposed by the decomp module makes it restrictive

SeanBryan51 commented on src/offline/cable_abort.F90:
```diff
@@ -20,55 +20,38 @@
+    CHARACTER(LEN=*), INTENT(IN) :: message !! Error message
+    CHARACTER(LEN=*), INTENT(IN), OPTIONAL :: file
+    INTEGER, INTENT(IN), OPTIONAL :: line
+    CHARACTER(5) :: line_string
```
I agree, I don't see why `file` and `line` shouldn't be compulsory - `OPTIONAL` only allowed that I could find and replace `abort` with `cable_abort` without having to specify the `file` and `line` arguments. It would be good to do this properly in a separate PR which addresses error handling (related https://github.com/CABLE-LSM/CABLE/issues/486).

SeanBryan51 commented on src/offline/cable_io_decomp.F90:
There should be no issues with declaring the decomposition objects into separate modules where they are used.

I suspect however that there will be decompositions which would be shared across applications when the in memory array and the NetCDF variable are the same shape. So in addition to separating the modules, I think there could be value in having a common decomposition object, where each application declares pointers which are associated with the common decomposition objects.

> how much does this change if we remove the capability to write out in different formats, and force an x-y grid format

If you enforce an x-y grid for all I/O then you should only be left with decompositions which involve `x_y` in the name.

SeanBryan51 committed:
src/offline/cable_io_decomp.F90: add int32 and real64 decompositions

SeanBryan51 committed:
src/offline/cable_mpimaster_stub.F90: update stub to accept mpi_grp_master

SeanBryan51 committed:
src/util/netcdf/cable_netcdf_decomp_util.F90: add missing allocations

SeanBryan51 committed:
src/util/netcdf/cable_netcdf.F90: add overloads for put_att and get_att

SeanBryan51 committed:
Move `range_abort` from cable_abort.F90 to cable_checks.F90

This change allows the `cable_abort_module` to be used in modules where
`cable_def_types_mod` or `cable_IO_vars_module` are a dependee of that
module as the removal of `range_abort` avoids introducing cyclic module
dependencies. The impact of this change is minimal as `range_abort` is
only called from the `cable_abort_module` in the code base.

SeanBryan51 committed:
src/offline/cable_io_decomp.F90: fix dim specification for 2D patch variables

SeanBryan51 committed:
src/util/netcdf/cable_netcdf.F90: add redef

SeanBryan51 committed:
src/util/netcdf/cable_netcdf.F90: Add iotype and mode arguments

SeanBryan51 committed:
src/util/netcdf/cable_netcdf.F90: make dim_names optional

SeanBryan51 committed:
src/util/netcdf/cable_netcdf.F90: Add inq_var_ndims procedure

SeanBryan51 committed:
src/offline/cable_parameters.F90: Add comment on init_local_structure_variables

SeanBryan51 committed:
tests/test_cable_netcdf.F90: specify NetCDF4 format in tests

SeanBryan51 committed:
Remove select rank as this is a fortran 2018 feature

SeanBryan51 committed:
src/util/netcdf/nf90/cable_netcdf_nf90.F90: Allocate  on decomposition initialisation instead of on distributed reads/writes

SeanBryan51 committed:
src/util/netcdf/pio/cable_netcdf_pio.F90: Add optional PIO namelist settings

SeanBryan51 committed:
Temporary commit: Add parallelio library built with openmpi/4.0.7 to enable DDT MPI degbugging

SeanBryan51 committed:
cable_netcdf_decomp_t should be abstract

SeanBryan51 committed:
Add base mode IO type

SeanBryan51 committed:
tests/fixtures.F90: Remove max string length constant

SeanBryan51 committed:
src/util/netcdf/cable_netcdf_decomp_util.F90: Remove max string length constant

SeanBryan51 committed:
src/util/netcdf/cable_netcdf.F90: Remove string length constants

SeanBryan51 committed:
src/util/netcdf/nf90/cable_netcdf_nf90.F90: Remove CABLE_NETCDF_MAX_RANK

SeanBryan51 committed:
src/util/netcdf/cable_netcdf.F90: Remove CABLE_NETCDF_MAX_RANK

SeanBryan51 committed:
src/util/netcdf/cable_netcdf.F90: Add API documentation

SeanBryan51 committed:
src/util/netcdf/cable_netcdf_stub_types.F90: Documentation

SeanBryan51 committed:
src/util/netcdf/pio/cable_netcdf_pio.F90: Documentation & code formatting

SeanBryan51 committed:
src/util/netcdf/nf90/cable_netcdf_nf90.F90: Documentation & code formatting

SeanBryan51 committed:
src/util/netcdf/nf90/cable_netcdf_nf90.F90: Remove redundant type component

SeanBryan51 committed:
src/util/netcdf/nf90/cable_netcdf_nf90.F90: Documentation

SeanBryan51 committed:
src/util/netcdf/pio/cable_netcdf_pio_stub.F90: Documentation & clean up

SeanBryan51 committed:
src/util/netcdf/cable_netcdf_internal.F90: Documentation

SeanBryan51 committed:
src/util/netcdf/cable_netcdf_decomp_util.F90: Documentation

SeanBryan51 commented:
Closing as this work is superseded by #706 and #712
